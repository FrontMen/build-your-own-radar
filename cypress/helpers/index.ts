export const dataCy = (value: string) => cy.get(`[data-cy=${value}]`);
export const dataTestId = (value: string) => cy.get(`[data-testid=${value}]`);

export const hexToRgb = (hex: string) => {
  const bigint = parseInt(hex.replace('#', ''), 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
}
