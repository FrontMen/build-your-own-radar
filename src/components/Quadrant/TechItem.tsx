import React from 'react';
import styled from 'styled-components';
import { Typography } from 'src/Theme/Typography';
import { MediaQueries } from 'src/Theme/Helpers';
import { Link } from 'react-router-dom';
import { IoIosArrowRoundForward } from 'react-icons/io';

const ListItem = styled.li`
  list-style: none;
  width: 100%;
  break-inside: avoid;
  overflow: hidden;

  @media ${MediaQueries.phablet} {
    min-width: 9em;
  }
`;

const Details = styled.div<{ isOpened: boolean }>`
  max-height: ${props => (props.isOpened ? '75vh' : '0px')};
  transition: max-height 0.2s ease-in-out;
  overflow: hidden;
  background-color: ${props => props.theme.pallet.light};
`;
const Content = styled.div`
  padding: ${props => `${props.theme.space[2]}px`};
  overflow: hidden;
`;

const Label = styled.div<{ highlighted: boolean }>`
  ${Typography.body};
  cursor: pointer;
  padding: ${props => props.theme.space[1] / 2}px;
  color: ${props => props.highlighted && props.theme.pallet.white};
  background-color: ${props => props.highlighted && props.theme.pallet.blue};
  font-weight: ${props => props.highlighted && 600};
`;

const DetailsLink = styled(Link)`
  flex-direction: row;
  align-items: center;
  margin-top: ${props => props.theme.space[2]}px;
  width: 100%;
  color: ${props => props.theme.pallet.blue};
  display: flex;
  font-weight: 400;
  font-size: 0.9em;
  text-decoration: none;
`;

const ArrowRightIcon = styled(IoIosArrowRoundForward)`
  height: 2em;
  width: 2em;
`;

export interface TechnologyProps {
  technology: Technology;
  highlighted: string | null;
  setHighlighted: (a: string | null) => void;
  selected: string | null;
  setSelected: (a: string | null) => void;
  quadrant: string;
}

// use Component here to do smart memo which is much more tricky to achieve with hooks
export class TechItem extends React.Component<TechnologyProps> {
  shouldComponentUpdate(nextProps: TechnologyProps) {
    const {
      highlighted: prevHighlighted,
      selected: prevActive,
      technology: { name: prevName },
    } = this.props;

    const {
      highlighted: nextHighlighted,
      selected: nextActive,
      technology: { name: nextName },
    } = nextProps;

    // updating if highlighted property changed
    if (
      (prevHighlighted === prevName && nextHighlighted !== nextName) ||
      (prevHighlighted !== prevName && nextHighlighted === nextName)
    ) {
      return true;
    }
    // updating if selected property changed
    if (
      (prevActive === prevName && nextActive !== nextName) ||
      (prevActive !== prevName && nextActive === nextName)
    ) {
      return true;
    }
    // otherwise not
    return false;
  }

  handleMouseOver = () => this.props.setHighlighted(this.props.technology.name);

  handleMouseOut = () => this.props.setHighlighted(null);

  handleClick = () => {
    const {
      setSelected,
      technology: { name },
      selected,
    } = this.props;
    setSelected(name === selected ? null : name);
  };

  render() {
    const {
      selected,
      highlighted,
      technology: { name, description },
    } = this.props;

    const quadrantSlug = this.props.quadrant.toLowerCase();
    const technologySlug = name.toLowerCase();
    return (
      <ListItem>
        <Label
          data-testid="label"
          onMouseOver={this.handleMouseOver}
          onMouseOut={this.handleMouseOut}
          onClick={this.handleClick}
          highlighted={highlighted === name}
        >
          {name}
        </Label>

        <Details data-testid="details" isOpened={selected === name}>
          <Content>
            <span dangerouslySetInnerHTML={{ __html: description }} />
            <DetailsLink to={`/${quadrantSlug}/${technologySlug}`}>
              {'More '}
              <ArrowRightIcon data-testid="arrow-right-icon" />
            </DetailsLink>
          </Content>
        </Details>
      </ListItem>
    );
  }
}
