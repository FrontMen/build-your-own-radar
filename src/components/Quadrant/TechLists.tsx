import React, { useMemo } from 'react';
import Groupby from 'lodash.groupby';
import { Section, Ring, Title, List, Dot, Tooltip } from './styled';
import { d3Config } from 'utils/d3-config';
import { TechItem } from './TechItem';
import { ringTooltip } from 'res/strings';

export interface TechnologiesListProps {
  technologies: Technology[];
  highlighted: string | null;
  setHighlighted: (a: string | null) => void;
  selected: string | null;
  setSelected: (a: string | null) => void;
  quadrant: string;
  color: string;
}

export const TechLists: React.FC<TechnologiesListProps> = ({
  technologies,
  highlighted,
  setHighlighted,
  selected,
  setSelected,
  quadrant,
  color,
}) => {
  const data = useMemo(() => Groupby(technologies, 'ring'), [technologies]);
  const availableRings = useMemo(
    () => Object.keys(d3Config.rings).filter(ringName => data[ringName]),
    [data],
  );

  return (
    <Section data-testid="tech-lists-section">
      {availableRings.map(ringName => (
        <Ring key={ringName}>
          <Title data-testid={`ring-title-${ringName}`}>
            <Dot color={color}>‚óè</Dot>
            {ringName}
            <Tooltip>
              ?<span>{ringTooltip[ringName as RingNamesType]}</span>
            </Tooltip>
          </Title>
          <List>
            {data[ringName] &&
              data[ringName].map(technology => (
                <TechItem
                  technology={technology}
                  key={technology.name}
                  selected={selected}
                  setSelected={setSelected}
                  highlighted={highlighted}
                  setHighlighted={setHighlighted}
                  quadrant={quadrant}
                />
              ))}
          </List>
        </Ring>
      ))}
    </Section>
  );
};
