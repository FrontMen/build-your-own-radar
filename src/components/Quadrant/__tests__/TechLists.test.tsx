import React from 'react';
import { mount } from 'enzyme';

import { TechLists, TechnologiesListProps } from '../TechLists';
import { TechItem } from '../TechItem';
import { mockData } from 'test/mockData';
import { AllProvidersWrapper, withAllProviders } from 'test/helpers';

const defaultProps: TechnologiesListProps = {
  highlighted: null,
  setHighlighted: jest.fn(),
  setSelected: jest.fn(),
  technologies: mockData,
  quadrant: 'tools',
  selected: null,
};

describe('TechLists', () => {
  test('it should matches the snapshot with default values', () => {
    const { container } = withAllProviders(<TechLists {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  test('it should pass down highlighted and setHighlighted props to list items', () => {
    const wrapper = mount(<TechLists {...defaultProps} />, {
      wrappingComponent: AllProvidersWrapper,
    });

    const label = wrapper.find(TechItem).first();

    expect(label.prop('highlighted')).toBe(defaultProps.highlighted);
    expect(label.prop('setHighlighted')).toBe(defaultProps.setHighlighted);
  });
});
