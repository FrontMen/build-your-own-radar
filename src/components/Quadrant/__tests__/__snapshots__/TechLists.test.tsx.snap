// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TechLists it should matches the snapshot with default values 1`] = `
.c4 {
  list-style: none;
  width: 100%;
  -webkit-break-inside: avoid;
  break-inside: avoid;
  overflow: hidden;
}

.c6 {
  max-height: 0px;
  -webkit-transition: max-height 0.2s ease-in-out;
  transition: max-height 0.2s ease-in-out;
  overflow: hidden;
}

.c7 {
  padding: 8px 0px;
  overflow: hidden;
}

.c5 {
  font-family: Roboto,san-serif;
  font-size: 1em;
  font-weight: 400;
  line-height: 1.65;
  cursor: pointer;
  padding: 0px 4px;
  border: 1px solid white;
}

.c8 {
  display: inline-block;
  font-weight: 400;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin-top: 8px;
  width: 100%;
  color: #526de4;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c1 {
  margin-bottom: 32px;
  width: 50%;
  min-width: 12rem;
  -webkit-break-inside: avoid;
  break-inside: avoid;
}

.c2 {
  font-family: Montserrat,san-serif;
  font-weight: 600;
  line-height: 1.15;
  font-size: 1;
  margin: 0;
  margin-bottom: 8px;
}

.c3 {
  padding-left: 0;
  margin: 0;
}

@media (min-width:481px) {
  .c4 {
    min-width: 9em;
  }
}

@media (min-width:481px) {
  .c2 {
    font-size: 1.5;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <h3
        class="c2"
      >
        Adopt
      </h3>
      <ul
        class="c3"
      >
        <li
          class="c4"
        >
          <div
            class="c5"
            data-testid="label"
          >
            Babel
          </div>
          <div
            class="c6"
            data-testid="details"
          >
            <div
              class="c7"
            >
              <span>
                This is the description. You can use basic html such as the 
                <strong>
                  strong tag to emphasise keywords and phrases
                </strong>
                 and insert
              </span>
              <a
                class="c8"
                href="/tools/babel"
              >
                Details &gt;
              </a>
            </div>
          </div>
        </li>
        <li
          class="c4"
        >
          <div
            class="c5"
            data-testid="label"
          >
            Spark
          </div>
          <div
            class="c6"
            data-testid="details"
          >
            <div
              class="c7"
            >
              <span>
                This is the description. You can use basic html such as the 
                <strong>
                  strong tag to emphasise keywords and phrases
                </strong>
                 and insert
              </span>
              <a
                class="c8"
                href="/tools/spark"
              >
                Details &gt;
              </a>
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="c1"
    >
      <h3
        class="c2"
      >
        Assess
      </h3>
      <ul
        class="c3"
      >
        <li
          class="c4"
        >
          <div
            class="c5"
            data-testid="label"
          >
            Electron
          </div>
          <div
            class="c6"
            data-testid="details"
          >
            <div
              class="c7"
            >
              <span>
                This is the description. You can use basic html such as the 
                <strong>
                  strong tag to emphasise keywords and phrases
                </strong>
                 and insert 
                <a
                  href="https://www.thoughtworks.com"
                >
                  anchor links to documentation and referance material
                </a>
                .
              </span>
              <a
                class="c8"
                href="/tools/electron"
              >
                Details &gt;
              </a>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
